tests/orderbook/test.js

```diff
@@ -0,0 +1,289 @@
+                "price": 10000000,
+        }
+
+        const expectAlice = {
+            "positions": {
+                "0": {
+                    "open_notional": 70000000,
+                    "size": 7000000000000000000,
+                    "unrealised_funding": null,
+                    "last_premium_fraction": 0,
+                    "liquidation_threshold": 7000000000000000000
+                }
+            },
+            "margins": {
+                "0": 29965000
+            }
+        }
+        const expectedBob = {
+            "positions": {
+                "0": {
+                    "open_notional": 70000000,
+                    "size": -7000000000000000000,
+                    "unrealised_funding": null,
+                    "last_premium_fraction": 0,
+                    "liquidation_threshold": -7000000000000000000
+                }
+            },
+            "margins": {
+                "0": 39965000
+            }
+        }
+
+        expect(evmState.order_map[longHash]).to.deep.include(expectOrder) // 1 partially filled open order left
+        expect(evmState.order_map[shortHash]).to.eq(undefined)
+        expect(evmState.trader_map[aliceAddress]).to.deep.eq(expectAlice)
+        expect(evmState.trader_map[bobAddress]).to.deep.eq(expectedBob)
+    });
+
+    it('Order cancel', async function () {
+        const { hash, order } = await placeOrder(alice, 2, 14)
+
+        tx = await orderBook.connect(alice).cancelOrder(order)
+        await tx.wait()
+
+        const evmState = await getEVMState()
+
+        expect(evmState.order_map[hash]).to.eq(undefined)
+    });
+
+    it('Order match error', async function () {
+        const {hash} = await placeOrder(charlie, 50, 12)
+        await placeOrder(bob, -10, 12)
+
+        const evmState = await getEVMState()
+        expect(evmState.order_map[hash]).to.eq(undefined) // should be deleted
+    });
+
+    it('Liquidate trader', async function () {
+        await addMargin(charlie, _1e6.mul(100))
+        await addMargin(alice, _1e6.mul(200))
+        await addMargin(bob, _1e6.mul(200))
+
+        await sleep(3)
+
+        // large position by charlie
+        await placeOrder(charlie, 49, 10)
+        await placeOrder(bob, -49, 10)
+
+        // reduce the price
+        await placeOrder(alice, 10, 8)
+        await placeOrder(bob, -10, 8)
+
+        // long order so that liquidation can run
+        const {hash} = await placeOrder(alice, 10, 8)
+
+        const expectedCharlie = {
+            "positions": {
+              "0": {
+                "open_notional": 360000000, // 49 - 10 - 3(from a previous order)
+                "size": 36000000000000000000,
+                "unrealised_funding": null,
+                "last_premium_fraction": 0,
+                "liquidation_threshold": 36000000000000000000
+              }
+            },
+            "margins": {
+              "0": 68555000
```

Comment:
how did we reach this value, my calculation says
```
100 - (46 * 10 * .05%) - (10 * 8 * 5%) = 95.77
```

Reply:
bob's short order(-8 * 10) is getting matched partially with the reset of charlie's order (3 quantity). Then Charlie's order gets liquidated twice - (3 + 10) - both charlie's orders with price 8

```
100 
- (49 * 10 * .05%) // total trade fee
- (13 * 8 * 5%) // liquidation penalty
- (13 * 2) // PnL because of liquidation
```